### PR TITLE
Expanding pow

### PR DESCRIPTION
**Pull request by William-TogetherScience**

### CODE REVIEW

File: ts-port/core/numbers.ts [hunk 1]
```diff
@@ -38,6 +41,20 @@ function igcd(x: number, y: number) {
     return x;
 }
 
+function ilcm(...args: any[]) {
```
Comment (William-TogetherScience) [inline ts-port/core/numbers.ts]: write a comment about what it does

File: ts-port/core/numbers.ts [hunk 2]
```diff
@@ -546,21 +603,21 @@ class Rational extends _Number_ {
...
                     const remfracpart = intpart * expt.q - expt.p;
                     const ratfracpart = new Rational(remfracpart, expt.q);
+                    const p1 = new Pow(new Integer(this.p), expt)
+                    const p2 = new Pow(new Integer(this.q), ratfracpart)
                     if (this.p !== 1) {
-                        // eslint-disable-next-line max-len
-                        return new Integer(this.p)._eval_power(expt).__mul__(new Integer(this.q))._eval_power(ratfracpart).__mul__(new Rational(1, this.q ** intpart, 1));
-                    }
-                    return new Integer(this.q)._eval_power(ratfracpart).__mul__(new Rational(1, this.q ** intpart, 1));
+                        return new Mul(true, true, p1, p2, new Rational(1, this.q ** intpart));
```
Comment (William-TogetherScience) [inline ts-port/core/numbers.ts:617]: double check (probably use decimal.js, check teams chat)

File: ts-port/core/numbers.ts [hunk 3]
```diff
@@ -611,6 +668,48 @@ class Rational extends _Number_ {
         return this.p === other.p && this.q === other.q;
     }
 
+    gcd(other: any) {
```
Comment (William-TogetherScience) [inline ts-port/core/numbers.ts:679]: write unit tests for unit tests (also for ilcm)

File: ts-port/core/numbers.ts [hunk 4]
```diff
@@ -611,6 +668,48 @@ class Rational extends _Number_ {
...
+            if (other === S.Zero) {
+                return other;
+            }
+            return new Rational(igcd(this.p, other.p), ilcm(this.q, other.q))
+        } else {
+            throw new Error("gcd not implemented for non rationals")
+        }
+    }
+
+    _Rrel(other: any, attr: any) {
```
Comment (William-TogetherScience) [inline ts-port/core/numbers.ts:690]: consider optimizing this away from late binding

File: ts-port/core/numbers.ts [hunk 5]
```diff
@@ -855,6 +954,24 @@ class Integer extends Rational {
         return result;
     }
 
+    __le__(other: any) {
```
Comment (William-TogetherScience) [inline ts-port/core/numbers.ts:965]: see if this deviates, and put in some comments

File: ts-port/core/power.ts [hunk 6]
```diff
@@ -128,23 +135,22 @@ export class Pow extends _Expr {
                 } else if (e === S.NegativeOne && !b) {
                     return S.ComplexInfinity;
                 } else if ((e.is_Symbol() && e.is_integer() ||
-                    e.is_Integer() && (b.is_Number() &&
-                    b.is_Mul() || b.is_Number())) && (e.is_extended_negative === true)) {
-                    if (e.is_even() || e.is_even()) {
+                    e.is_Integer()) && ((b.is_number() &&
```
Comment (William-TogetherScience) [inline ts-port/core/power.ts]: why did this get changed??

File: ts-port/core/power.ts [hunk 7]
```diff
@@ -168,6 +174,126 @@ export class Pow extends _Expr {
         return [b, e];
     }
 
+    _eval_is_finite() {
```
Comment (William-TogetherScience) [inline ts-port/core/power.ts:177]: why did I need this (and eval_is_zero)

File: ts-port/core/power.ts [hunk 8]
```diff
@@ -168,6 +174,126 @@ export class Pow extends _Expr {
...
+        if (other.is_integer()) {
+            s = S.One
+        } else if (typeof e.is_extended_real() !== "undefined") {
+            // HELPER FUNCTIONS
+            function _half(e: any) {
+                if (e.q === 2) {
+                    return true;
+                }
+                const [n, d] = e.as_numer_denom();
+                // PLACEHOLDER FOR IS_EQ : TO-DO UPDATE THIS !!!!!
```
Comment (William-TogetherScience) [inline ts-port/core/power.ts:248]: do this

File: ts-port/core/relational.ts [hunk 9]
```diff
@@ -0,0 +1,192 @@
+/*
+Notes and changes (by WB):
+- Currently implementing only enough for POW to work well (extremely barebones,
+  will port more later when it becomes necessary)
```
Comment (William-TogetherScience) [inline ts-port/core/relational.ts:4]: fix header comment

File: ts-port/core/relational.ts [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+            const diff = new Add(true, true, lhs, new Mul(true, true, rhs, S.NegativeOne));
+            if (diff !== S.NaN) {
+                const rv = diff.is_extended_nonnegative()
+                if (typeof rv !== "undefined") {
+                    return rv
+                }
+            }
+        }
+    }
+}
```
Comment (William-TogetherScience) [inline ts-port/core/relational.ts]: line break

File: ts-port/core/utility.ts [hunk 11]
```diff
@@ -204,6 +204,16 @@ class Util {
         const match = s.match(/^(.+?)\s+(.+?)\s+(.*)/);
         return match?.slice(1);
     }
+
+    static abs(val: any) {
+        if (val.is_Symbol()) {
```
Comment (William-TogetherScience) [inline ts-port/core/utility.ts]: this is questionable - see what sympy does

**Review by gunnarmein-ts — APPROVED** [review]

**Pull request merged**